src/mesh/mod.rs

```diff
@@ -796,6 +822,35 @@ fn remove_orphan_vertices_and_normals(
     (faces_renumbered, vertices_reduced, normals_reduced)
 }
 
+pub fn calculate_smooth_normals_from_components(
+    vertices: &[Point3<f32>],
+    faces: &[Face],
+    vertex_to_face_topology: &[SmallVec<[u32; topology::MAX_INLINE_NEIGHBOR_COUNT]>],
+) -> Vec<Vector3<f32>> {
+    let vertex_count = vertices.len();
+
+    let mut normals: Vec<Vector3<f32>> = Vec::with_capacity(vertex_count);
+
+    for shared_face_indices in vertex_to_face_topology {
+        let shared_faces_count_f32 = shared_face_indices.len() as f32;
+        let mut normal: Vector3<f32> = Vector3::zeros();
+        for face_index in shared_face_indices {
+            match faces[cast_usize(*face_index)] {
+                Face::Triangle(face) => {
+                    normal += geometry::compute_triangle_normal(
```

Comment:
I'm thinking now it would be better to calculate all face normals upfront and then use the lookup table for these calculations. That way we reduce calls of this function to 1/3. Which is worse, memory allocation or heavier computation?

Reply:
Depends. If the calculation is linear (it is in this case), I would say memory allocation is worse. If the calculation is quadratic and memory allocation helps, I tend to prefer allocation unless I can somehow prove the data is very small.